Add a configurable order number generator

tvdeyen opened the pull request:
The default order number generator was extracted into a class.
Provide your own class that takes an options hash and returns a
string when called for `#generate`.

    Spree::Config.order_number_generator = TrueNumberGenerator

    class TrueNumberGenerator
      def initialize(options = {})
      end

      def generate
        '42'
      end
    end

jhawthorn commented on core/app/models/spree/order/number_generator.rb:
```diff
@@ -0,0 +1,28 @@
+        # Use the random number if no other order exists with it.
+        if Spree::Order.exists?(number: random)
+          # If over half of all possible options are taken add another digit.
+          length += 1 if Spree::Order.count > (10**length / 2)
```
This is a good catch by hound. I think there's an issue here between `length`  being a instance method and being assigned here as a variable.

tvdeyen replied:
Oh, actually a good catch. So, the only solution I see is to use a `attr_accessor` for `length` then, right?

jhawthorn commented:
I like this. It would be even better if we could share the code to generate Order numbers, Shipment numbers, and Payment numbers

tvdeyen commented:
> share the code to generate Order numbers, Shipment numbers, and Payment numbers

@jhawthorn I also thought about this, but I think we should not make the assumption that all numbers generate the same way. 

The only idea coming to my mind is the ask the `NumberGenerator` for `#order_number`, `#shipment_number`, `#payment_number` and so on. WDYT?

tvdeyen commented:
@jhawthorn addressed the issues. 

Still think this is valuable as order number generator on its own. Shipments and friends get their number through `permalinks.rb` and are not that visible to customers as the order number. 

Let's address this - totally valid concern - in a separate PR, ok?

jhawthorn commented on core/app/models/spree/app_configuration.rb:
```diff
@@ -365,6 +365,17 @@ def tax_adjuster_class
+    # @api experimental
+    attr_writer :order_number_generator
+    def order_number_generator
+      @order_number_generator ||= Spree::Order::NumberGenerator
```
One thought: maybe this should be an instance of `NumberGenerator` rather than the class. This would allow developers to specify a length, prefix, or letters without needing to write their own class

tvdeyen commented:
@jhawthorn good idea. I needed to remove the possibility for passing `options` into `Spree::Order#generate_order_number`, though. But I think this was barely used anyway.
